### ISSUE TITLE
Dependency-check return null with folder link alias 

### ISSUE THREAD

**Issue by b82** [dependency-check/DependencyCheck#1947]
I've noticed that the plugin fails its scan and return `null` when it tries to scan a node_module **folder link**. It tries to go inside the folder but doesn't find a real folder, so it fails and returns error:

```
WARN] An unexpected error occurred during analysis of '/project/npm-shrinkwrap.json' (Node.js Package Analyzer): null
[ERROR]
java.lang.NullPointerException: null
	at org.glassfish.json.JsonObjectBuilderImpl$JsonObjectImpl.getString(JsonObjectBuilderImpl.java:199)
	at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.processDependencies(NodePackageAnalyzer.java:253)
	at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.processDependencies(NodePackageAnalyzer.java:259)
	at org.owasp.dependencycheck.analyzer.NodePackageAnalyzer.analyzeDependency(NodePackageAnalyzer.java:226)
	at org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze(AbstractAnalyzer.java:138)
	at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:88)
	at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:37)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[INFO] Finished Node.js Package Analyzer (0 seconds)
```

Monorepo or big projects has lots of submodules with their dependencies and folder links helps `npm i` for better performances, but this represents a problem for the plugin.
I don't know if it is a bug or a possible feature, but I think it's a good idea to let you know about this behaviour. Is it possible for dependency-check to follow the link and scan the correct folder (and complete correctly the scan) or is it a project self limit if this plugin is used for it?

Thank you very much

**Comment by b82** [comment]
It should be a bug btw, is there a way to change the issue label?

**Comment by thib3113** [comment]
Hello @b82 , sorry for the long delay, I just see your issue, and maybe having something similar .

If I correctly understand, your project include some other projects ? in sub_folders ?

Are the other package.json checked by dependency-check ?
Did you agree that skipping the scan of the sub modules, when scanning the root  `npm-shrinkwrap.json` can be a solution ?

Maybe you can share your `npm-shrinkwrap.json` ? 

Else, are your submodules respecting this schema (in `npm-shrinkwrap.json`) ?
Is version the same schema ? `file:name_of_the_submodule`
```
{
    "fake_submodule": {
      "version": "file:fake_submodule",
      "requires": {
        "braces": "^1.8.5"
      },
      "dependencies": {
        "braces": {
          "version": "1.8.5",
          "resolved": "https://registry.npmjs.org/braces/-/braces-1.8.5.tgz",
          "integrity": "sha1-uneWLhLf+WnWt2cR6RS3N4V79qc=",
          "requires": {
            "expand-range": "^1.8.1",
            "preserve": "^0.2.0",
            "repeat-element": "^1.1.2"
          }
        }
      }
    }
}
```

**Comment by Jmaharman** [comment]
I'm running into this issue as well. Any linked package, such as below, causes the error:

```
"node_modules/utils": {
    "resolved": "packages/utils",
    "link": true
},
```

Does anyone know how to avoid the error? It scans all the dependencies, but the exception is always logged and makes you wonder whether it missed some dependencies.

**Comment by rturner-edjuster** [comment]
> I'm running into this issue as well. Any linked package, such as below, causes the error:
> 
> ```
> "node_modules/utils": {
>     "resolved": "packages/utils",
>     "link": true
> },
> ```
> 
> Does anyone know how to avoid the error? It scans all the dependencies, but the exception is always logged and makes you wonder whether it missed some dependencies.

I've been digging into this as well today, as I'm struggling with the same NPE as result. I think the code needs updating to recognize a "link" entry, and simply skip it without trying to parse the version attribute (which doesn't get added by NPM to the link entry). I'm not an expert on the NPM package_lock file, but examining https://github.com/jeremylong/DependencyCheck/blob/main/core/src/main/java/org/owasp/dependencycheck/analyzer/NodePackageAnalyzer.java suggests that a partial fix was added to `shouldSkipDependency`, but an additional change is also needed in `processDependencies` where it blindly assumes that every entry that is a JsonObject has a version key.

Unfortunately because it's a crash, I can't see a quick and easy way to work around it, unless there is a way to tell DependencyCheck to not report a failure in such a case.

A potential "fix" might be something like:
```java
                if (entry.getValue() instanceof JsonObject) {
                    jo = (JsonObject) entry.getValue();

                    // Ignore/skip linked entries
                    if (jo.getBoolean("link", false) {
                        continue;
                    }

                    version = jo.getString("version");
                    optional = jo.getBoolean("optional", false);
                    isDev = jo.getBoolean("dev", false);
                } else {
                    version = ((JsonString) entry.getValue()).getString();
                }
```

**Comment by jeremylong** [comment]
@rturner-edjuster any chance you could open a PR with the suggested fix?

**Comment by rturner-edjuster** [comment]
@jeremylong Sure -- might take me a few hours as I'll need to get all set up with the project, etc, and learn all the project semantics, and how to get an automated test in for it, etc. But I'll see what I can do this evening.

**Comment by rturner-edjuster** [comment]
@jeremylong PR is up and waiting for approval to run workflows, etc. Seems to fix the issue in my environment locally, and the test I added passes with the fix, and crashes without. Let me know if you need any other changes done, or if I haven't completed things in compliance with the project guidelines.

**Comment by angus-88** [comment]
When is this going to be release? We have had to disable Dependency checks currently while we wait

**Comment by jeremylong** [comment]
This should be fixed in 8.0.0.
